federation/executor.go

```diff
@@ -88,3 +91,100 @@ func NewExecutor(ctx context.Context, executors map[string]ExecutorClient) (*Exe
 	}, nil
 
 }
+
+func (e *Executor) runOnService(ctx context.Context, service string, typName string, keys []interface{}, kind string, selectionSet *graphql.SelectionSet) (map[string]interface{}, error) {
+	// Execute query ons pecified service
+	schema := e.Executors[service]
+	bytes, err := schema.Execute(ctx, &graphql.Query{
+		Kind:         kind,
+		SelectionSet: selectionSet,
+	})
+	if err != nil {
+		return nil, oops.Wrapf(err, "execute remotely")
+	}
+	// Unmarshal json from results
+	var res interface{}
+	if err := json.Unmarshal(bytes, &res); err != nil {
+		return nil, oops.Wrapf(err, "unmarshal res")
+	}
+	result, ok := res.(map[string]interface{})
+	if !ok {
+		return nil, fmt.Errorf("executor res not a map[string]interface{}")
+	}
+	return result, nil
+}
+
+func (e *Executor) execute(ctx context.Context, p *Plan, keys []interface{}) (interface{}, error) {
+	res := map[string]interface{}{}
+
+	// Executes that part of the plan (the subquery) on one of the federated gqlservers
+	if p.Service != gatewayCoordinatorServiceName {
+		var err error
+		res, err = e.runOnService(ctx, p.Service, p.Type, keys, p.Kind, p.SelectionSet)
+		if err != nil {
+			return nil, oops.Wrapf(err, "run on service")
+		}
+	}
+
+	g, ctx := errgroup.WithContext(ctx)
+	var resMu sync.Mutex
+
+	// For every nested query in the plan, execute it on the specified service and stitch
+	// the results into a response
+	for _, currentSubPlan := range p.After {
+		subPlan := currentSubPlan
+		var subPlanMetaData pathSubqueryMetadata
+		if p.Service == gatewayCoordinatorServiceName {
+			subPlanMetaData.keys = nil // On the root query there are no specified keys
+			subPlanMetaData.results = []map[string]interface{}{res}
+		}
+
+		g.Go(func() error {
+			// Execute the subquery on the specified service
+			results, err := e.execute(ctx, subPlan, subPlanMetaData.keys)
+			if err != nil {
+				return fmt.Errorf("executing sub plan: %v", err)
+			}
+
+			// Acquire mutex lock before modifying results
+			resMu.Lock()
+			defer resMu.Unlock()
+			for _, subPlanResults := range subPlanMetaData.results {
+				result, ok := results.(map[string]interface{})
+				if !ok {
+					return fmt.Errorf("result is not an object: %v", result)
+				}
+				for k, v := range result {
+					subPlanResults[k] = v
+				}
+			}
+			return nil
+		})
+	}
+
+	if err := g.Wait(); err != nil {
```

Comment:
Should we combine `subPlanResults` to `res` somewhere?

Reply:
Since `subPlanMetaData.results` points to res, we are directly updating it already. We also acquire the mutex lock right before this section so that its thread-safe.